crates/sui-indexer-alt/src/handlers/obj_info.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use anyhow::{anyhow, Result};
+use diesel_async::RunQueryDsl;
+use sui_types::{base_types::ObjectID, full_checkpoint_content::CheckpointData, object::Owner};
+
+use crate::{
+    db,
+    models::objects::{StoredObjInfo, StoredOwnerKind},
+    pipeline::{concurrent::Handler, Processor},
+    schema::obj_info,
+};
+
+pub struct ObjInfo;
+
+impl Processor for ObjInfo {
+    const NAME: &'static str = "obj_info";
+    type Value = StoredObjInfo;
+
+    fn process(&self, checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
+        let cp_sequence_number = checkpoint.checkpoint_summary.sequence_number as i64;
+        let checkpoint_input_objects = checkpoint.checkpoint_input_objects();
+        let latest_live_output_objects = checkpoint
+            .latest_live_output_objects()
+            .into_iter()
+            .map(|o| (o.id(), o))
+            .collect::<BTreeMap<_, _>>();
+        let mut values: BTreeMap<ObjectID, Self::Value> = BTreeMap::new();
+        for object_id in checkpoint_input_objects.keys() {
+            if !latest_live_output_objects.contains_key(object_id) {
+                // If an input object is not in the latest live output objects, it must have been deleted
+                // or wrapped in this checkpoint. We keep an entry for it in the table.
+                // This is necessary when we query objects and iterating over them, so that we don't
```

Comment:
I am not following here, can you elaborate on the query that needs to use deleted / wrapped entries?
also looks like there is no "marker column" marking if the object ID is deleted/wrapped, would that be problematic, for example if a query asking for 50 IDs ended up getting some deleted objects, and as a result end graphql response has < 50 results?

Reply:
> can you elaborate on the query that needs to use deleted / wrapped entries?

All queries need to track deleted / wrapped objects, because they will start by finding candidates that meet the ownership/type criteria, and they will follow that up by checking whether there is some later version of the object that supersedes the candidate (in which case we discard it), and an object should be considered superseded if it has been deleted or wrapped.

> would that be problematic, for example if a query asking for 50 IDs ended up getting some deleted objects, and as a result end graphql response has < 50 results?

We have this problem today with `object_history` and `objects_snapshot` (or `wal_obj_types` and `sum_obj_types`) and we solve it by pushing the `LIMIT` as close into the query as possible. E.g. if we need to find 50 matching objects, we:

- Select 50 matching candidates from the summary/snapshot table, after removing objects where there's a newer version in the history/wal table.
- Select 50 matching candidates from the history/wal table, after removing objects where there's a newer version in the history/wal table.
- Union both of these together, order by object ID and again limit it to 50 results.

With `obj_info`, it works similarly, it's just that the summary and WAL tables are combined into one:

- Select 50 matching candidates from `obj_info` after removing objects that have been updated at later checkpoints in `obj_info`.
- Map these candidates to their latest versions in the checkpoint they are being viewed at (This is something that @lxfind and I have been struggling with a little bit on how to make this efficient).

Reply:
This issue is avoidable though, and maybe we should give it a try.
We would need to make sure that for each transfer, we insert two entries into this table, and make then distinguishable. Then we could naturally get rid of the out-of-dated entries in the join